### PR TITLE
feat: b2b user module

### LINKED ISSUE #2: As a B2B User, I should be able to view other users in my B2B Account
A B2B user should be able to view other users in the same B2B account from a link, titled "Users" in the My Account section.

Upon clicking the "Users" link, user should be redirected to a "/my-account/b2b-users" page. (Use the my-account/order-history) for reference

The page should display a table list of other b2b users, allow space for icons and include the following fields
* Email
* First Name
* Last Name
* Role
* Status

Above the table, a search input component should be displayed with the placeholder message "Search user by first name, last name, or email address"

Create a new React Hook to fetch and search for b2bAccountUsers

Hook should use GQL query b2bAccountUsers

https://www.figma.com/file/bKJuIwUx6VXmubHZo4rCBq/B2B?node-id=19-688&t=buaa5Uu85x81VveX-0

### LINKED ISSUE #3: As a B2B Admin User, I should be able to add users to my B2B Account
As an Admin user viewing the b2b-users page, I want to add new B2B users into my account.

An "Add User" button should be displayed above the b2b users list.

When the button is clicked from Desktop View
* Display a form inline with fields, email address, firstname, lastname and radio buttons for Role
* Display a Cancel and Save Button
* Use existing Snackbar to show Fail or Success Status 
* All fields are required
* Same Form input will be used later for "Edit" functionality, make re-usable by allowing the form fields and Save callback function to be passed as Props 

UX: https://www.figma.com/file/bKJuIwUx6VXmubHZo4rCBq/B2B?node-id=121-4725&t=buaa5Uu85x81VveX-0

When the button is clicked from Mobile View
* Display a Full width Modal
* Modal should be form with fields, email address, firstname, lastname and radio buttons for Role
* Display a Cancel and Save Button
*  Use existing Snackbar to show Fail or Success Status 
* All fields are required

UX: https://www.figma.com/file/bKJuIwUx6VXmubHZo4rCBq/B2B?node-id=141-7916&t=buaa5Uu85x81VveX-0

Create a React Hook that calls GQL Mutation createCustomerB2bAccountUser

### LINKED ISSUE #4: As a B2B Admin User, I should be able to edit users in my B2B Account
B2B Admin users should be able to edit the email address, first name, last name, role and current status of other b2b users in the account

### LINKED ISSUE #5: As a B2B Admin User, I should be able to delete users from my B2B Account

### PR DESCRIPTION
**Pull request by bsatignitiv**

### CODE REVIEW

File: components/common/SearchBar/SearchBar.tsx
Comment (kibo-chandradeeptalaha) [inline components/common/SearchBar/SearchBar.tsx:1]: revert this change. It may effect the header search

**Review by kibo-sushant — APPROVED** [review]

File: components/dialogs/UserFormDialog/UserFormDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import React from 'react'
+
+import { useTranslation } from 'next-i18next'
+
+import KiboDialog from '@/components/common/KiboDialog/KiboDialog'
```
Comment (kibo-chandradeeptalaha) [inline components/dialogs/UserFormDialog/UserFormDialog.tsx]: Fix the imports. Should use @/components/common

File: components/dialogs/UserFormDialog/UserFormDialog.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+const Template: ComponentStory<typeof UserFormDialog> = ({ ...args }) => (
+  <UserFormDialog {...args} />
+)
+
+// Common
+export const Common = Template.bind({})
+Common.args = {
+  isEditMode: false,
+  b2BUser: undefined,
+  onSave: (data: B2BUser) => console.log(data),
```
Comment (kibo-chandradeeptalaha) [inline components/dialogs/UserFormDialog/UserFormDialog.stories.tsx]: Can we move this to argTypes ?

File: components/my-account/User/UserForm/UserForm.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+const Template: ComponentStory<typeof UserForm> = (args) => <UserForm {...args} />
+
+// My Account
+export const Common = Template.bind({})
+Common.args = {
+  isEditMode: false,
+  b2BUser: undefined,
+  onClose: () => console.log('form closed'),
+  onSave: (formValues) => console.log(formValues),
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserForm/UserForm.stories.tsx]: Move these to argTypes

File: components/my-account/User/UserForm/UserForm.tsx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                  !isUserFormInDialog &&
+                  classes.cancelButtonInDesktopEditMode) ||
+                (isDesktopView && !isUserFormInDialog && classes.cancelButtonInDesktop) ||
+                classes.cancelButtonInMobile
+              }
+            >
+              {isDesktopEditView && !isUserFormInDialog ? <ClearIcon /> : t('cancel')}
+            </Button>
+
+            <Button
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserForm/UserForm.tsx]: use LoadingButton component. You can find reference in CartTemplate.tsx

File: components/my-account/User/UserForm/UserForm.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+  } = useForm({
+    defaultValues: { role: 'Admin', emailAddress: '', firstName: '', lastName: '', isActive: true },
+    resolver: yupResolver(userSchema),
+  })
+  const onSubmit = async () => {
+    if (isLoading) return
+    setLoading(true)
+    const formValues = getValues()
+    if (isEditMode) {
+      onSave(formValues, b2BUser)
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserForm/UserForm.tsx]: Aren't we using await ?

File: components/my-account/User/UserTable/UserTable.tsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+              <TableCell>{t('last-name-or-sur-name')}</TableCell>
+            </>
+          )}
+          <TableCell>{t('role')}</TableCell>
+          {mdScreen && <TableCell>{t('status')}</TableCell>}
+          <TableCell></TableCell>
+        </TableRow>
+      </TableHead>
+      <TableBody>
+        {!b2bUsers?.length ? (
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserTable/UserTable.tsx]: We can use Caption to show this.

Comment (kibo-chandradeeptalaha) [reply components/my-account/User/UserTable/UserTable.tsx]: You can check Quick Order Table in our PR for reference

**Review by kibo-chandradeeptalaha — CHANGES_REQUESTED** [review]

**Review by kibo-chandradeeptalaha — CHANGES_REQUESTED** [review]
Please check

File: components/my-account/User/UserTable/UserTable.tsx (hunk 6 above)
Comment (bsatignitiv) [reply components/my-account/User/UserTable/UserTable.tsx]: Done

File: components/my-account/User/UserForm/UserForm.tsx (hunk 4 above)
Comment (bsatignitiv) [reply components/my-account/User/UserForm/UserForm.tsx]: Done

File: components/my-account/User/UserForm/UserForm.tsx (hunk 5 above)
Comment (bsatignitiv) [reply components/my-account/User/UserForm/UserForm.tsx]: Added await

File: components/my-account/User/UserForm/UserForm.stories.tsx (hunk 3 above)
Comment (bsatignitiv) [reply components/my-account/User/UserForm/UserForm.stories.tsx]: Done

File: components/dialogs/UserFormDialog/UserFormDialog.stories.tsx (hunk 2 above)
Comment (bsatignitiv) [reply components/dialogs/UserFormDialog/UserFormDialog.stories.tsx]: Done

File: components/dialogs/UserFormDialog/UserFormDialog.tsx (hunk 1 above)
Comment (bsatignitiv) [reply components/dialogs/UserFormDialog/UserFormDialog.tsx]: Done

File: components/common/SearchBar/SearchBar.tsx
Comment (bsatignitiv) [reply components/common/SearchBar/SearchBar.tsx:1]: @kibo-chandradeeptalaha Keeping my changes as per the discussion with you

File: components/my-account/User/UserTable/UserTable.tsx [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  TableRow,
+  Typography,
+  styled,
+  useMediaQuery,
+  useTheme,
+} from '@mui/material'
+import { useTranslation } from 'next-i18next'
+
+import UserForm from '../UserForm/UserForm'
+import UserFormDialog from '@/components/dialogs/UserFormDialog/UserFormDialog'
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserTable/UserTable.tsx]: Fix import

File: components/my-account/User/UserTable/UserTable.tsx [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                {b2bUser?.emailAddress}
+              </TableCell>
+              {mdScreen && (
+                <>
+                  <TableCell sx={{ flex: 1 }}>{b2bUser?.firstName}</TableCell>
+                  <TableCell sx={{ flex: 1 }}>{b2bUser?.lastName}</TableCell>
+                </>
+              )}
+              <TableCell sx={{ flex: 1 }}>
+                {(b2bUser?.roles && b2bUser?.roles.length && b2bUser?.roles[0]?.roleName) || 'N/A'}
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserTable/UserTable.tsx]: Can't we get these values from getters ?

File: components/my-account/User/UserTable/UserTable.tsx (hunk 7 above)
Comment (bsatignitiv) [reply components/my-account/User/UserTable/UserTable.tsx]: Done

File: components/my-account/User/UserTable/UserTable.tsx (hunk 8 above)
Comment (bsatignitiv) [reply components/my-account/User/UserTable/UserTable.tsx]: I have changed 
b2bUser?.roles && b2bUser?.roles.length && b2bUser?.roles[0]?.roleName

to 

b2bUser?.roles?.[0]?.roleName || 'N/A'

which is more readable.

File: components/my-account/User/UserForm/UserForm.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  })
+
+  it('should reset the form and call onClose', async () => {
+    render(<Common {...WithProps.args} onSave={onSave} onClose={onClose} />)
+
+    // Access the cancel button element
+    const cancelButton = screen.getByTestId('reset-button')
+
+    // Simulate a click on the cancel button
+    fireEvent.click(cancelButton)
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserForm/UserForm.spec.tsx:70]: use userEvent.

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx [hunk 10]
```diff
@@ -1,18 +1,309 @@
...
+    <Button
+      variant="primary"
+      disabled={isUserFormOpen}
+      onClick={onClick}
+      disableElevation
+      id="formOpenButton"
+      startIcon={<AddCircleOutlineIcon />}
+      sx={sx}
+    >
+      <span>{t('add-user')}</span>
```
Comment (kibo-chandradeeptalaha) [inline components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Remove the span

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx [hunk 11]
```diff
@@ -1,18 +1,309 @@
...
+    const variables = buildCreateCustomerB2bUserParams({ user, values: formValues })
+    const createUserResponse = await createCustomerB2bUser.mutateAsync({
+      ...variables,
+    })
+    if (createUserResponse?.userId) {
+      addRoleToB2bUser(createUserResponse, formValues)
+    }
+  }
+
+  const handleUpdateUser = async (formValues: B2BUserInput, b2BUser?: B2BUser | undefined) => {
```
Comment (kibo-chandradeeptalaha) [inline components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: b2BUser?: B2BUser | undefined

If we are using ?, no need to explicitly provide undefined again

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx [hunk 12]
```diff
@@ -1,18 +1,309 @@
...
+    <Grid>
+      <Grid item style={{ marginTop: '10px', marginBottom: '20px' }}>
+        <BackButtonLink aria-label={t('my-account')} href="/my-account">
+          <ChevronLeftIcon />
+          {mdScreen && <Typography variant="body1">{t('my-account')}</Typography>}
+        </BackButtonLink>
+        <Box sx={UsersTemplateStyle.heading}>
+          <Typography
+            variant={mdScreen ? 'h1' : 'h2'}
+            sx={{ textAlign: { xs: 'center' }, marginTop: { xs: '-3.6rem', md: 0 } }}
```
Comment (kibo-chandradeeptalaha) [inline components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Instead of negative margin, use padding

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx [hunk 13]
```diff
@@ -1,18 +1,309 @@
...
+          >
+            {t('users')}
+          </Typography>
+        </Box>
+        <Grid container>
+          <Grid item xs={12} md={12}>
+            <AddUserButton
```
Comment (kibo-chandradeeptalaha) [inline components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Don't create a separate component for this

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx (hunk 10 above)
Comment (bsatignitiv) [reply components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Done

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx (hunk 11 above)
Comment (bsatignitiv) [reply components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Done

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx (hunk 12 above)
Comment (bsatignitiv) [reply components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Done

File: components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx (hunk 13 above)
Comment (bsatignitiv) [reply components/page-templates/B2B/UsersTemplate/UsersTemplate.tsx]: Done

**Review by kibo-chandradeeptalaha — APPROVED** [review]
